### ISSUE TITLE
[FP] identity_op in front of if

### ISSUE THREAD

**Issue by hellow554** [rust-lang/rust-clippy#8724]
### Summary

The lint triggers, although the suggested code procudes a syntax error.

Problem is most likely, that the lint only views the if expression and not the whole one, e.g. it's a sum of (at least) three operands, where the first one is the identity (`0` in this case). But the expression is invalid without that, because that's how `if` is parsed in rust right now.

### Lint Name

identity_op

### Reproducer

I tried this code:

```rust
pub fn decide(a: bool, b: bool) -> u32 {
    0 + if a { 1 } else { 2 } + if b { 3 } else { 5 }
}
```

I saw this happen:

```
warning: the operation is ineffective. Consider reducing it to `if a { 1 } else { 2 }`
 [--> src/lib.rs:2:5
](https://play.rust-lang.org/#)  |
2 |     0 + if a { 1 } else { 2 } + if b { 3 } else { 5 }
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(clippy::identity_op)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#identity_op
```

I expected to see this happen:

### Version

```text
rustc 1.62.0-nightly (311e2683e 2022-04-18)
binary: rustc
commit-hash: 311e2683e1bad87715b1558f7900e294d24ce491
commit-date: 2022-04-18
host: x86_64-unknown-linux-gnu
release: 1.62.0-nightly
LLVM version: 14.0.1
```

### Additional Labels

@rustbot label: I-suggestions-causes-error

**Comment by tamaroning** [comment]
This kind of false positives is true for other expressions (e.g. `match`).
And the folowing code compiles.
```
pub fn decide(a: bool, b: bool) -> u32 {
    (if a { 1 } else { 2 }) + if b { 3 } else { 5 }
}
```

Which is better, not triggering the lint, or triggering the lint and suggesting using `(` `)`?
Anyone has some idea?

**Comment by tamaroning** [comment]
@xFrednet
Thanks for your advice!

BTW, is there any way to make sub expression not be visited?
If I use `LateLintPass::check_expr` against `let _ = 0 + if b { 1 } else { 2 } + if b { 3 } else { 4 };`,
both `0 + if b { 1 } else { 2 } + if b { 3 } else { 4 }` and `0 + if b { 1 } else { 2 }` are visited.
I want to make `0 + if b { 1 } else { 2 }` not to be visted because this form should trigger the lint, though the entire expression shoud not trigger the lint.

**Comment by tamaroning** [comment]
I found this can be done by `walk_expr`
